src/Comparison.dfy

```diff
@@ -0,0 +1,93 @@
+// RUN: %dafny /compile:0 "%s"
+
+/*******************************************************************************
+*  Copyright by the contributors to the Dafny Project
+*  SPDX-License-Identifier: MIT 
+*******************************************************************************/
+
+include "Wrappers.dfy"
+include "Relations.dfy"
+module Comparison {
+
+  import opened Wrappers
+  import opened Relations
+
+  datatype Cmp = Lt | Eq | Gt {
+    function method Flip(): Cmp {
+      match this
+        case Lt => Gt
+        case Eq => Eq
+        case Gt => Lt
+    }
+
+    const Le? := this != Gt
+    const Ge? := this != Lt
+
+    static function method ComputeTransitivity(c0: Cmp, c1: Cmp): Option<Cmp> {
+      match (c0, c1)
+        case (Lt, Lt) => Some(Lt)
+        case (Lt, Eq) => Some(Lt)
+        case (Lt, Gt) => None
+        case (Eq, Lt) => Some(Lt)
+        case (Eq, Eq) => Some(Eq)
+        case (Eq, Gt) => Some(Gt)
+        case (Gt, Lt) => None
+        case (Gt, Eq) => Some(Gt)
+        case (Gt, Gt) => Some(Gt)
+    }
+  }
+   datatype Comparison<!T> = Comparison(cmp: (T, T) -> Cmp) {
+    function method Compare(t0: T, t1: T): Cmp {
+      cmp(t0, t1)
+    }
+
+     predicate Complete??(t0: T, t1: T) {
+      cmp(t0, t1) == cmp(t1, t0).Flip()
```

Comment:
It threw me a bit that I didn't recognize this isn't one of the standard properties of total orderings, until I realized that "Transitive" really meant "cmp(...).Le? is a transitive binary relation" and similarly for "Reflexive". Is "complete" a standard term for this property of this kind of comparator?

I expect at the point where we try to connect `Relations` to this version, we'll have a lemma that proves that `C.Valid(s) ==> Relations.TotalOrdering((t1, t2) => C.cmp(t1, t2).Le?)`